src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorSourceUpdater.cs

```diff
@@ -48,59 +56,332 @@ public async Task<IUpdaterResult> IsApplicableAsync(IUpgradeContext context, Imm
-            _diagnostics = allDiagnostics
+            var ret = allDiagnostics
                 .Where(d => d.Location.IsInSource &&
-                       inputs.Select(i => $"{i.Source.FilePath}.cs").Contains(d.Location.SourceTree.FilePath) &&
+                       targetFilePaths.Contains(d.Location.SourceTree.FilePath) &&
                        _codeFixProviders.Any(f => f.FixableDiagnosticIds.Contains(d.Id)))
-                .Distinct(comparer);
+                .Distinct(diagnosticsComparer);
+
+            return ret;
+        }
+
+        private static Project GetProjectWithGeneratedCode(IProject project, IEnumerable<RazorCodeDocument> documents)
+        {
+            var projectWithGeneratedCode = project.GetRoslynProject();
+            foreach (var document in documents)
+            {
+                var filePath = GetGeneratedFilePath(document);
+                projectWithGeneratedCode = projectWithGeneratedCode.AddDocument(Path.GetFileName(filePath), document.GetCSharpDocument().GeneratedCode, null, filePath).Project;
+            }
+
+            return projectWithGeneratedCode;
+        }
+
+        private async Task<Project> TryFixDiagnosticAsync(Diagnostic diagnostic, Document document)
+        {
+            CodeAction? fixAction = null;
+            var context = new CodeFixContext(document, diagnostic, (action, _) => fixAction = action, CancellationToken.None);
+            var fixProvider = _codeFixProviders.FirstOrDefault(p => p.FixableDiagnosticIds.Contains(diagnostic.Id));
+
+            if (fixProvider is not null)
+            {
+                await fixProvider.RegisterCodeFixesAsync(context).ConfigureAwait(false);
+            }
+
+            // fixAction may not be null if the code fixer is applied.
+#pragma warning disable CA1508 // Avoid dead conditional code
+            if (fixAction is null)
+#pragma warning restore CA1508 // Avoid dead conditional code
+            {
+                _logger.LogWarning("No code fix found for {DiagnosticId}", diagnostic.Id);
+                return document.Project;
+            }
+
+            var applyOperation = (await fixAction.GetOperationsAsync(CancellationToken.None).ConfigureAwait(false)).OfType<ApplyChangesOperation>().FirstOrDefault();
+
+            if (applyOperation is null)
+            {
+                _logger.LogWarning("Code fix could not be applied for {DiagnosticId}", diagnostic.Id);
+                return document.Project;
+            }
 
-            _logger.LogDebug("Identified {DiagnosticCount} fixable diagnostics in Razor files in project {ProjectName}", _diagnostics.Count(), project.Name);
+            return applyOperation.ChangedSolution.GetProject(document.Project.Id) ?? throw new InvalidOperationException("Updated project not found");
         }
 
         private static IAsyncEnumerable<DiagnosticAnalyzer> GetApplicableAnalyzersAsync(IEnumerable<DiagnosticAnalyzer> analyzers, IProject project) =>
```

Comment:
This class is very large. Can it be broken up into smaller services? I don't see any tests in this PR (were they already merged in?), but the smaller services will make the testing much easier

Reply:
Yea, that's a good idea. I'm adding tests in a follow-up PR, but you're right that will be simpler if this is broken up. I'll refactor it a bit.

Reply:
+1 on making this class smaller.

Reply:
Yep, the class has been shrunk :)